Add more details to enhanced nav assessment switcher

miguelaenlle opened the pull request:
Closes #11555.

- Added assessment sets/module grouping, AIDs, open issue counts, and group assignment indicators to the enhanced navigation assessment switcher. 
- The assessment switcher loads content via an `htmx` `GET` request to the `assessments_switcher` endpoint instead of generating it directly in `AssessmentNavigation.html.ts` from `resLocals.assessments`

Assessments grouped by set:

![image](https://github.com/user-attachments/assets/60962662-4cc9-4a6b-bc88-0c208ec33a01)

Assessments grouped by module:

![image](https://github.com/user-attachments/assets/a7465e70-4ace-4990-b543-6ba3f726f801)

Mobile view:

![image](https://github.com/user-attachments/assets/f0035c09-2d49-484a-b7c2-9ba7a4a3fe42)

miguelaenlle commented on apps/prairielearn/src/components/AssessmentNavigation.html.ts:
```diff
@@ -39,22 +28,23 @@ export function AssessmentNavigation({
+        <div
+          id="assessmentNavigationDropdownContent"
+          style="max-height: 50vh"
+          class="overflow-auto py-2"
```
This needs an empty state.

miguelaenlle replied:
Test with a small number of assessments, too.

miguelaenlle commented on apps/prairielearn/src/pages/assessmentsSwitcher/assessmentsSwitcher.ts:
```diff
@@ -0,0 +1,42 @@
+
+    res.send(
+      AssessmentSwitcher({
+        urlPrefix: res.locals.urlPrefix,
```
Should this be moved down?

miguelaenlle commented on apps/prairielearn/src/pages/assessmentsSwitcher/assessmentsSwitcher.sql:
```diff
@@ -0,0 +1,67 @@
+-- BLOCK select_assessment_dropdown_items_data
```
This query is a simplified version of [this query](https://github.com/PrairieLearn/PrairieLearn/blob/master/apps/prairielearn/src/pages/instructorAssessments/instructorAssessments.sql#L1-L80) in `instructorAssessments.sql`.

mylesw27 replied:
Since these are so similar, can we use the same query in both places?

jonatanschroeder commented:
Any chance of including the AID in the title itself as well?

miguelaenlle commented:
> Any chance of including the AID in the title itself as well?

Hi @jonatanschroeder . I'd be happy to add it, and will add the AID to the title itself shortly.

miguelaenlle commented:
> Any chance of including the AID in the title itself as well?

@jonatanschroeder , I added the AID to the dropdown:

<img width="1710" alt="image" src="https://github.com/user-attachments/assets/8e2c5064-673c-4edf-8e46-68787569b62b" />

Let me know your thoughts on this design.

mylesw27 commented on apps/prairielearn/src/components/AssessmentNavigation.html.ts:
```diff
@@ -40,22 +28,25 @@ export function AssessmentNavigation({
+        hx-trigger="mouseover once, focus once, show.bs.dropdown once delay:200ms"
+        hx-target="#assessmentNavigationDropdownContent"
       >
         <span class="h6 mb-0 me-1 overflow-hidden text-truncate">${assessment.title}</span>
```
Sorry, I know this is not part of this PR, but still related. Do we want to use something bigger than `.h6` here? This feels so small and IMO gets lost in the page. Feel free to ignore if this has been discussed previously and I missed it. Wondering if something like `.h4` would look better here?

Something like this:
![Screenshot 2025-03-24 at 5 07 43 PM](https://github.com/user-attachments/assets/2ff779e5-9c3d-4986-bd0f-618c45ba271f)

miguelaenlle replied:
In #11279, @nwalters512 suggested reducing the assessment selector size some time ago, which is why I used the `.h6` size

https://github.com/PrairieLearn/PrairieLearn/pull/11279#issuecomment-2660163421

> Re: the assessment chooser that's visible at the top when navigating within a single assessment:
> The assessment chooser seems really large. Can we use a more typical font size for it?

mylesw27 replied:
Sounds good! I am good with `.h6` then since it has already been discussed.

mylesw27 commented on apps/prairielearn/src/components/AssessmentNavigation.html.ts:
```diff
@@ -40,22 +28,25 @@ export function AssessmentNavigation({
+        hx-get="/pl/assessments_switcher/course_instance/${courseInstanceId}/assessment/${assessment.id}${subPage
+          ? `?subPage=${subPage}`
+          : ''}"
+        hx-trigger="mouseover once, focus once, show.bs.dropdown once delay:200ms"
```
What is the reasoning behind using `htmx` instead of using the list of assessments we already have on `res.locals.assessments`?

miguelaenlle replied:
I used `htmx` instead of `res.locals.assessments` to retrieve assessment set/module names and the `start_new_assessment_group` column. 

To include assessment set/module information while using `res.locals.assessments`, I could alternatively:
- Fetch all assessment sets/modules for the course instance in a separate query, then match that information with the `res.locals.assessments` in TypeScript
- Adjust the `select_assessments` query to include that information (after all, it performs a `LEFT JOIN` on the `assessment_sets` table already)

Let me know if you think the current implementation is okay or if I should use the `res.locals.assessments` list instead.

mylesw27 replied:
Makes sense, I was just curious. The current implementation looks good.

mylesw27 commented on apps/prairielearn/src/pages/assessmentsSwitcher/assessmentsSwitcher.html.ts:
```diff
@@ -0,0 +1,100 @@
+          href="${plainUrlPrefix}/course_instance/${courseInstanceId}/instructor/assessment/${assessmentDropdownItemData.id}/${targetSubPage ??
+          ''}"
+        >
+          <div class="d-flex align-items-center" style="width: 50px; min-width: 50px;">
+            <span class="badge color-${assessmentDropdownItemData.assessment_set.color} mb-auto">
+              ${assessmentDropdownItemData.label}
+            </span>
+          </div>
```
![Screenshot 2025-03-24 at 4 53 02 PM](https://github.com/user-attachments/assets/bae1a7ed-6900-4631-837c-f7a5bf6d6117)
It does not take many characters for the label to run into the assessment title & id. Do we want to do something to handle overflow here?

miguelaenlle replied:
Thank you for pointing this out, @mylesw27 . I'll implement overflow handling here 👍

mylesw27 reviewed (commented):
This looks pretty good! I had some initial questions below. Apologies if any of these have been discussed elsewhere and I missed it.

miguelaenlle commented:
> Since these are so similar, can we use the same query in both places?
Considering that the `select_assessment_dropdown_items_data` query doesn't retrieve any more information than the query in `instructorAssessments.sql`, I agree that it's better if we use the same query in both places. I'll implement this change

miguelaenlle commented on apps/prairielearn/src/pages/assessmentsSwitcher/assessmentsSwitcher.html.ts:
```diff
@@ -0,0 +1,102 @@
+} from '../../lib/db-types.js';
+import { idsEqual } from '../../lib/id.js';
+
+export const AssessmentDropdownItemDataSchema = AssessmentSchema.extend({
```
Remove this, and discontinue its use.

miguelaenlle commented on apps/prairielearn/src/pages/assessmentsSwitcher/assessmentsSwitcher.html.ts:
```diff
@@ -0,0 +1,102 @@
+type AssessmentDropdownItemData = z.infer<typeof AssessmentDropdownItemDataSchema>;
+
+export function AssessmentSwitcher({
+  assessmentDropdownItemsData,
```
Rename this to `assessmentRows` (or similar)

miguelaenlle commented on apps/prairielearn/src/pages/assessmentsSwitcher/assessmentsSwitcher.sql:
```diff
@@ -0,0 +1,67 @@
+-- BLOCK select_assessment_dropdown_items_data
```
Remove this file

miguelaenlle commented on apps/prairielearn/src/pages/assessmentsSwitcher/assessmentsSwitcher.ts:
```diff
@@ -0,0 +1,56 @@
+    });
+
+    const assessmentDropdownItemsData = await queryRows(
+      sql.select_assessment_dropdown_items_data,
```
Use the correct query

miguelaenlle reviewed (commented):
Notes from my self-review.

miguelaenlle commented on apps/prairielearn/src/lib/assessment.sql:
```diff
@@ -1391,3 +1391,83 @@ FROM
   LEFT JOIN assessments AS a ON (a.id = ai.assessment_id)
   LEFT JOIN course_instances AS ci ON (ci.id = a.course_instance_id);
+
+-- BLOCK select_assessments_for_course_instance_grouped
```
This is the `select_assessments` query from `instructorAssessments.sql`.

miguelaenlle commented on apps/prairielearn/src/pages/instructorAssessments/instructorAssessments.ts:
```diff
@@ -130,7 +118,7 @@ router.get(
       // update assessment statistics if needed
       await updateAssessmentStatisticsForCourseInstance(res.locals.course_instance.id);
 
-      const cursor = await queryCursor(sql.select_assessments, {
```
While the `selectAssessmentsForCourseInstanceGroupedCursor` function is used only here, I added that function to `lib/assessment.ts` so that the `select_assessments` query only has to stay in one file, `lib/assessment.sql`.

mylesw27 replied:
Looks good to me!

miguelaenlle commented:
> Since these are so similar, can we use the same query in both places?

To reuse the `select_assessments` query from `instructorAssessments.sql` in `assessmentsSwitcher.ts`, I moved it to `select_assessments_for_course_instance_grouped` in `lib/assessment.sql`: 

https://github.com/PrairieLearn/PrairieLearn/blob/1b84cd1c6141c4c1000772084675f8480a2b537b/apps/prairielearn/src/lib/assessment.sql#L1395-L1473

This is then used by `selectAssessmentsForCourseInstanceGrouped`, which is then called in `instructorAssessments.ts` and `assessmentsSwitcher.ts`.

https://github.com/PrairieLearn/PrairieLearn/blob/1b84cd1c6141c4c1000772084675f8480a2b537b/apps/prairielearn/src/lib/assessment.ts#L573-L599

Let me know your thoughts on this implementation, @mylesw27 .

mylesw27 reviewed (commented):
This is looking good to me. I believe @nwalters512 still wants to take a look so I am not approving yet but this all looks good on my end.

miguelaenlle commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -201,3 +201,73 @@ body:has(div.app-container) {
+  flex-direction: column;
+}
+
+.assessment-heading {
```
Place `#assessment-switcher-container` in front of all the assessment switcher styles

miguelaenlle commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -201,3 +201,73 @@ body:has(div.app-container) {
+.assessment-row {
+  width: 100%;
+  display: grid;
+  grid-template-columns: minmax(75px, auto) 1fr 1fr;
```
Comment rationale for minmax

miguelaenlle commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -201,3 +201,73 @@ body:has(div.app-container) {
+  border-radius: 5px;
+}
+
+.assessment-row.selected a {
```
Comment why this is separate; remove one of the color: white statements if possible

miguelaenlle commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -201,3 +201,73 @@ body:has(div.app-container) {
+  display: none;
+}
+
+.bottom-label {
```
Is there a better way to handle the two labels? Maybe ::nth-child?

miguelaenlle commented on apps/prairielearn/src/components/AssessmentNavigation.html.ts:
```diff
@@ -1,63 +1,58 @@
-              </a>
-            `;
-          })}
+    <button
```
There's a white background behind this button. Ensure that this isn't a regression.

miguelaenlle commented on apps/prairielearn/src/pages/assessmentsSwitcher/assessmentsSwitcher.html.ts:
```diff
@@ -0,0 +1,85 @@
+                </div>
+              `
+            : ''}
+          <div class="assessment-row ${isActive ? 'selected' : ''}" aria-label="Link to assessment">
```
Remove the aria-label, it is useless

miguelaenlle commented on apps/prairielearn/src/pages/assessmentsSwitcher/assessmentsSwitcher.ts:
```diff
@@ -0,0 +1,48 @@
+  asyncHandler(async (req, res) => {
+    const subPage = req.query.subPage as NavSubPage | undefined;
+
+    // Target subpage for the assessment dropdown links
```
Ensure that this still works

miguelaenlle reviewed (commented):
Comments from my self-review.

nwalters512 commented:
A few notes! Though I'll acknowledge that I'm getting into "just having fun with it" territory and might be spending more time than I should on a feature that most instructors or going to be in and out of without paying too much attention to it. That said, I like building nice UIs, so I'll share my thoughts anyways.

If you think any of these aren't worth it, by all means tell me 🙏 

## Long assessment names

In the hypothetical case of a really long assessment badge, things get kind of weird:

<img width="835" alt="Screenshot 2025-05-08 at 14 48 55" src="https://github.com/user-attachments/assets/02b6a0d2-3183-49a8-902e-112d98e292df" />

I was wondering about taking some inspiration from how GitHub handles long labels, specifically doing truncation with an elipsis:

<img width="327" alt="Screenshot 2025-05-08 at 14 50 10" src="https://github.com/user-attachments/assets/b7c8da78-59dc-4100-adda-746cd261c35d" />

This seems to basically just be the following CSS:

```css
overflow: hidden;
text-overflow: ellipsis;
white-space: nowrap;
```

I wonder if we could do the same here? Here's a mockup:

<img width="823" alt="Screenshot 2025-05-08 at 14 52 05" src="https://github.com/user-attachments/assets/4e383b01-2d3e-4d08-93c7-81b5c0d32184" />

We could potentially do the same thing for long assessment directory names too. Here's what that could look like (truncating at the start, not the end):

<img width="828" alt="Screenshot 2025-05-08 at 15 05 23" src="https://github.com/user-attachments/assets/45ea79a3-3393-4f72-b361-ab467016b717" />

And the CSS:

```
overflow: hidden;
direction: rtl;
text-overflow: ellipsis;
```

## Reduced visual weight of directory names

I'm wondering if we should apply `text-muted` to the assessment directories? Here's how that would look on desktop and mobile:

<img width="824" alt="Screenshot 2025-05-08 at 14 55 21" src="https://github.com/user-attachments/assets/99a58af8-13b9-4b16-9d1d-65171e117159" />
<img width="409" alt="Screenshot 2025-05-08 at 14 56 02" src="https://github.com/user-attachments/assets/de092ffb-4011-45f6-a247-d90e2909a9b7" />

## Hover/focus state

We could maybe include a nice background when a row is hovered or has focus within it. Here's a mockup:

<img width="828" alt="Screenshot 2025-05-08 at 15 05 23" src="https://github.com/user-attachments/assets/d47159f3-4e28-4998-a3d8-e30bdb1c49e0" />

Here's the CSS I used for that:

```css
.assessment-row:not(.selected):is(:hover, :focus-within) {
  background-color: var(--bs-secondary-bg);
  border-radius: 5px;
}
```

What do you think?

If we go that route, we probably want the cells to use flexbox to vertically center their content. E.g. in the screenshot, note how "C4" isn't vertically centered.

miguelaenlle commented:
> Long assessment names
- I agree with using ellipsis to handle long assessment badges, and will implement this. 

> Reduced visual weight of directory names
- `text-muted` would certainly differentiate the assessment name from the directory name and place less emphasis on it. I'll add that

> Hover/focus state
- I initially implemented this functionality but removed it since the entire row isn't clickable. I think that this is a good addition though, since we don't set `cursor` to `pointer` when hovering over a row.

miguelaenlle commented on apps/prairielearn/src/lib/assessment.sql:
```diff
@@ -1824,13 +1824,11 @@ FROM
   assessments AS a
   JOIN course_instances AS ci ON (ci.id = a.course_instance_id)
   LEFT JOIN assessment_sets AS aset ON (aset.id = a.assessment_set_id)
-  LEFT JOIN LATERAL authz_assessment (a.id, $authz_data, $req_date, ci.display_timezone) AS aa ON TRUE
```
I made this change to ensure that this query remains the same as `select_assessments` in `/instructorAssessments/instructorAssessments.sql` (which is where it comes from)

miguelaenlle commented:
@nwalters512 , I implemented all the suggested changes:

- Long assessment labels are now truncated:

![image](https://github.com/user-attachments/assets/ee6b5189-038d-4c98-8b56-a7b52b2efc0e)

- Long directory names are also now truncated:

![image](https://github.com/user-attachments/assets/f0941fb6-87e9-401d-bf66-982e780ba97e)

- Applied `text-muted` to the TIDs to reduce their visual weight

- Implemented a hover background for the assessment rows and centered all assessment row content with flexbox:

![image](https://github.com/user-attachments/assets/01611753-ef85-432c-8a8f-6a1b8fe9244a)

nwalters512 commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -201,3 +201,93 @@ body:has(div.app-container) {
   padding: 0.5rem 1rem;
 }
+
+#assessment-switcher-container {
+  display: flex;
+  flex-direction: column;
+}
```
Nit: we typically prefer to use Bootstrap styles/classes wherever possible. In this case, it'd be more conventional to use Bootstrap's utility classes inline (i.e. `d-flex` and `flex-column`) directly on the element itself. Ditto for many of the things below (layout, padding, etc.). I know there's no great way to do grid config inline like that, so I'm happy to keep that here in CSS.

miguelaenlle replied:
I reimplemented most of the CSS-based styling in Bootstrap. This significantly simplified the CSS for the assessment switcher:

https://github.com/PrairieLearn/PrairieLearn/blob/7cba06b604869babe9f5091728f3ccafc7a26f5e/apps/prairielearn/assets/stylesheets/pageLayout.css#L205-L242

A few notes:
- I kept the `assessment-switcher-container` ID in both the HTML and CSS to ensure that assessment switcher styles only apply to assessment switcher elements
- I left the `assessment-row` padding and margin in CSS since it changes based on the user's device size:

https://github.com/PrairieLearn/PrairieLearn/blob/7cba06b604869babe9f5091728f3ccafc7a26f5e/apps/prairielearn/assets/stylesheets/pageLayout.css#L205C1-L211

https://github.com/PrairieLearn/PrairieLearn/blob/7cba06b604869babe9f5091728f3ccafc7a26f5e/apps/prairielearn/assets/stylesheets/pageLayout.css#L218-L226

nwalters512 replied:
Excellent! Yeah, these styles look much cleaner now. I left a few more notes about this in my latest review.

nwalters512 commented on apps/prairielearn/src/lib/assessment.ts:
FYI: both you and @SethPoulsen are competing to see who will land this shared assessment selection function first 🙂 he's doing the same in https://github.com/PrairieLearn/PrairieLearn/pull/11770

I actually prefer your placement; since we're selecting assessments, it belongs in `assessments.ts`.

nwalters512 commented on apps/prairielearn/src/server.ts:
```diff
@@ -492,6 +492,13 @@ export async function initExpress(): Promise<Express> {
     ],
   );
+  app.use(
+    '/pl/assessments_switcher/course_instance/:course_instance_id(\\d+)/assessment/:assessment_id(\\d+)?',
```
Nit: can we use the same `/pl/navbar` namespace as the above routes? I'd propose this (you'll have to update the URL elsewhere):

```suggestion
    '/pl/navbar/course_instance/:course_instance_id(\\d+)/assessment/:assessment_id(\\d+)/switcher',
```

Aside: can you say more about why the assessment ID bit of the route was listed as optional? AFAIK we should always be rendering this menu in the context of a specific assessment. If it doesn't have to be optional, let's fix that and update the relevant types on this page.

miguelaenlle replied:
> `/pl/navbar` namespace
Sounds good; I'll make this change 👍 

> Assessment ID listed as optional
The assessment ID was listed as optional since I initially assumed that it could be called if no assessment ID was selected. If no assessment ID was specified, the HTML for the assessment switcher would be returned, just without an assessment selected. 

Since this menu is always rendered in the context of a specific assessment, I'll set the assessment ID to be required.

nwalters512 replied:
Makes sense, thanks!

nwalters512 commented on apps/prairielearn/src/pages/assessmentsSwitcher/assessmentsSwitcher.ts:
```diff
@@ -0,0 +1,46 @@
+
+import { AssessmentSwitcher } from './assessmentsSwitcher.html.js';
+
+const router = Router({
+  mergeParams: true, // Ensures that assessmentsSwitcher can retrieve req.locals.course_instance and req.params.assessment_id from the parent router
+});
```
Nit: we already do this on most pages, so I don't think this is notable enough to call out here.

```suggestion
const router = Router({ mergeParams: true });
```

nwalters512 commented on apps/prairielearn/src/pages/assessmentsSwitcher/assessmentsSwitcher.html.ts:
```diff
@@ -0,0 +1,84 @@
+  targetSubPage?: NavSubPage;
+}) {
+  if (assessmentRows.length === 0) {
+    return html` <p class="m-0">No assessments</p> `.toString();
```
```suggestion
    return html`<p class="m-0">No assessments</p>`.toString();
```

nwalters512 reviewed (commented):
A few more comments, then I think this is good to go 🙏

SethPoulsen commented:
Hey @miguelaenlle I felt bad about leaving you a merge conflict by landing my PR, so I went ahead and resolved the merge conflict for you, hope that's ok.

SethPoulsen commented:
I just realized your choice of function name `selectAssessmentsForCourseInstance` is probably better than my choice of `selectAssessment`, feel free to change it back in this PR if you want.

miguelaenlle commented:
> Hey @miguelaenlle I felt bad about leaving you a merge conflict by landing my PR, so I went ahead and resolved the merge conflict for you, hope that's ok.

Thank you for resolving the merge conflict @SethPoulsen ! I appreciate it.

miguelaenlle commented:
> I just realized your choice of function name `selectAssessmentsForCourseInstance` is probably better than my choice of `selectAssessment`, feel free to change it back in this PR if you want.

I'll leave the `selectAssessments` name since `course_instance_id` is explicitly shown as one of the parameters of the function, implying that `selectAssessments` filters by it:

https://github.com/PrairieLearn/PrairieLearn/blob/7cba06b604869babe9f5091728f3ccafc7a26f5e/apps/prairielearn/src/pages/assessmentsSwitcher/assessmentsSwitcher.ts#L26-L28

nwalters512 commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -201,3 +201,42 @@ body:has(div.app-container) {
+  display: grid;
+  /* Use minmax(0, 1fr) to prevent content overflow */
+  grid-template-columns: 75px minmax(0, 1fr) minmax(0, 1fr);
+  gap: 10px;
```
Did you consider using one of the [`gap-*` utilities](https://getbootstrap.com/docs/5.3/utilities/spacing/#gap) from Bootstrap instead of this?

miguelaenlle replied:
I did, but implemented `gap/padding/margin` in the CSS file instead of with Bootstrap since many other CSS classes in `pageLayout.css` (e.g. `side-nav-link`, `side-nav-header`) have `padding, margin`, etc. specified in the CSS file. 

However, for consistency with the rest of the codebase and to simplify this implementation, I moved the `gap/padding/margin` styles for elements related to the assessment nav switcher into Bootstrap. 

@nwalters512 , would you recommend opening a separate PR to move the other styles in `pageLayout.css` that can reasonably be moved to inline Bootstrap?

nwalters512 commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -201,3 +201,42 @@ body:has(div.app-container) {
+    /* The title takes up the entire first row. The label and TID take up the second row. */
+    grid-template-columns: auto 1fr;
+    grid-template-rows: auto auto;
+    column-gap: 0.5rem;
+    row-gap: 0;
```
Ditto re: using utilities from Bootstrap.

nwalters512 commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -201,3 +201,42 @@ body:has(div.app-container) {
+    grid-template-rows: auto auto;
+    column-gap: 0.5rem;
+    row-gap: 0;
+    padding: 0.5rem;
+    margin-top: 0.25rem;
```
Typically we'd also reach for Bootstrap utilities in this situation too. `0.25rem` top margin would map to `mt-1`, `0.5rem` padding would map to `p-2`. There are responsive variants too, e.g. `mt-md-1`.

nwalters512 reviewed (approved):
Just a few final notes about how the styles are implemented. I won't block on these though. Excellent work here!
